src/render/src/lib.rs

```diff
@@ -146,70 +140,259 @@ pub mod shade;
 pub mod macros;
 */
 
+use std::collections::VecDeque;
+use core::{CommandQueue, QueueType, Surface, Swapchain, Device as CoreDevice};
+use core::pool::{CommandPool, CommandPoolCreateFlags};
+use core::format::RenderFormat;
+use memory::Typed;
+
+struct Queue<B: Backend, C> {
+    inner: CommandQueue<B, C>,
+    command_pool_receiver: encoder::CommandPoolReceiver<B, C>,
+    command_pool_sender: encoder::CommandPoolSender<B, C>,
+}
+
+impl<B: Backend, C> Queue<B, C> {
+    fn new(inner: CommandQueue<B, C>) -> Self {
+        let (command_pool_sender, command_pool_receiver) =
+            encoder::command_pool_channel();
+        Queue { inner, command_pool_sender, command_pool_receiver }
+    }
+
+    fn acquire_encoder_scope(&mut self) -> encoder::Scope<B, C> {
+        let initial_capacity = 4;
+        let flags = CommandPoolCreateFlags::empty();
+        let pool = self.command_pool_receiver.try_recv()
+            .map(|mut recycled| {
+                recycled.reset();
+                recycled
+            })
+            .unwrap_or_else(|_| {
+                CommandPool::from_queue(&self.inner, initial_capacity, flags)
+            });
+        encoder::Scope::new(pool, self.command_pool_sender.clone())
+    }
+}
+
+/* TODO
 pub(crate) enum Queue<B: Backend> {
     General(CommandQueue<B, core::General>),
     Graphics(CommandQueue<B, core::Graphics>),
 }
+*/
 
-pub struct Gpu<B: Backend> {
+// TODO: could be named Instance too ?
+pub struct Context<B: Backend> {
+    surface: B::Surface,
     device: Device<B>,
-    queue: Queue<B>,
+    queue: Queue<B, core::Graphics>,
```

Comment:
no compute yet?

Reply:
Not yet, it was simpler to write. I thought about having two methods to `acquire_general/graphics_scope` but both can fail (even graphics since there is no downcasting). This would not be very convenient to use. I also thought about having an enum inside the `Encoder` but it seemed annoying to match all the time in the implementation.

Reply:
Can we lift the `C` generic to the `Gpu` level (in `gfx_render`) and have 2 kind of constructors: one with compute (so, `Generic`) and other without (so, `Graphics`). More often than not, the client knows if they need to run compute.

Reply:
Works great :)